### DISCUSSION TITLE
Model solves with longer timesteps, but not shorter ones.

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#23497] [Q&A General] [answered]
Hello
I have a real conundrum which seems counter-intiutive to how I though MOOSE works.

I have an input file which runs quickly and efficiently when default timestep dt = 1s.  It takes about a minute to do 1000 timesteps with all timesteps = 1s, i.e. not cutbacks are needed.

If I change the defualt timestep to dt = 0.1s, then after a few steps it starts cutting back the timestep to even shorter times, struggles to solve and ultimately reaches the minimum timestep value and fails to complete the solve.

If I make dt = 10s, it solves quickly with no cutbacks.  The solution is pretty much the same as for dt = 1s, but coarser as expected.

I had always thought that if a solve was struggling, making timesteps shorter would help with this.  But here exactly the opposite is happening!

Has anybody else seen this anomalous behaviour?  Any ideas on what is causing it and how to fix it?

Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Could you please attach the convergence log for the 0.1s case and the regular case too

Thank you
Guillaume

**Reply by EdSheltonUKAEA** [reply]
Thanks for the quick reply!  Can you tell where to find or how to generate the convergence log?  I'll try to get this done later today - but more likely tomorrow.

**Reply by GiudGiud** [reply]
it's what appears on the console when you run the input basically. If you can attach them rather than paste them

**Reply by EdSheltonUKAEA** [reply]
Hi 
These are the outputs for the start of each run.  I hope I have attached them OK.

You will see that with dt=1s, the solve gets to 60s with no problems or cutbacks.
With dt=0.1s, it starts having to cut back at 1.4s.


[ConvergenceLog1_dt=1,0s_NoCutback_SolveSucceeds.txt](https://github.com/idaholab/moose/files/10796371/ConvergenceLog1_dt.1.0s_NoCutback_SolveSucceeds.txt)
[ConvergenceLog2_dt=0,1s_Cutback_SolveFails.txt](https://github.com/idaholab/moose/files/10796372/ConvergenceLog2_dt.0.1s_Cutback_SolveFails.txt)

**Reply by GiudGiud** [reply]
It seems the line search dies with the 0.1s timesteps.
Do you mind turning it off? With line_search = none in Executioner

With regards to finding the culprit, can you please paste your executioner block here
also can you turn on variable residual output. one can use `show_var_residual_norms` in Debug block
https://mooseframework.inl.gov//syntax/Debug/index.html

**Reply by EdSheltonUKAEA** [reply]
Thanks!  I will try this.
Here's the Executioner block before making the changes you suggest:

```
[Executioner]
  scheme = bdf2
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -ksp_type -snes_type -pc_factor_shift_type'
  petsc_options_value = 'lu bcgs newtonls NONZERO'
  automatic_scaling = true
  compute_scaling_once = false
  nl_abs_tol = 1e-15 # This should avoid solve failures when close to steady-state
  nl_rel_tol = 1e-5
  dt = 0.1 # s 
  timestep_tolerance = 1e-3 # s
  end_time = 10 # s
[]
```

**Reply by EdSheltonUKAEA** [reply]
Attached now is the convergence log with "line_search = none" included.
The lines stating "Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH ..." are no longer there (as expected I suppose), but the solve still fails after 1.4 s.
There's also an extract of the log from 1.3 s with the Debug residuals command included.

[ConvergenceLog3_dt=0,1s_LineSearchOff_SolveFails.txt](https://github.com/idaholab/moose/files/10802222/ConvergenceLog3_dt.0.1s_LineSearchOff_SolveFails.txt)
[ConvergenceLog3a_dt=0,1s_LineSearchOff_DebugResidualsOnFrom_t=1,3s_SolveFails.txt](https://github.com/idaholab/moose/files/10802224/ConvergenceLog3a_dt.0.1s_LineSearchOff_DebugResidualsOnFrom_t.1.3s_SolveFails.txt)

**Reply by GiudGiud** [reply]
Can you turn off re-scaling on every iteration
  compute_scaling_once = false
by making it true. 

This is changing the goal for reducing residual in each variable on every time step .

If we look around the failure, you have a lot of variables and some of them are very well converged already. 
I think the key here will be to come up with a good scaling strategy. 
You can either try manual scaling (in Variables block) or try to group variables in scaling_groups, in the executioner parameters
https://mooseframework.inl.gov/source/executioners/Steady.html#613ee2f4-a2b2-400a-953d-62f040f096dc

```
Time Step 14, time = 1.30625, dt = 0.00625

Performing automatic scaling calculation

 0 Nonlinear |R| = 6.900904e-06
    |residual|_2 of individual variables:
                   C_Hm:     3.72346e-20
                   C_Htr1_1: 2.43051e-06
                   C_Htr1_2: 8.49783e-08
                   C_Htr1_3: 6.43852e-06
                   C_Htr1_4: 5.03338e-07
                   C_Htr1_5: 2.78968e-09
                   C_Htr1_6: 2.17528e-12
                   C_Dm:     8.74736e-24
                   C_Dtr1_1: 4.90283e-10
                   C_Dtr1_2: 1.70592e-11
                   C_Dtr1_3: 1.2877e-09
                   C_Dtr1_4: 1.00667e-10
                   C_Dtr1_5: 5.57936e-13
                   C_Dtr1_6: 4.35055e-16
```

**Reply by EdSheltonUKAEA** [reply]
Changing compute_scaling_once = true doesn't seem to make any difference.  It still breaks down after 1.3 s.
The comment about scaling agrees with what colleagues here have suggested.
I'll need to understand more about how this works, so will take a look at the links.
Are they any other resources on the MOOSE website or elsewhere which explain scaling calculations and how to apply manual scaling?  It would be really helpful to see a simple example which shows what calculations are being made by the solver, and how scaling helps to make these easier (or even possible) to calculate.

**Reply by GiudGiud** [reply]
There is not good documentation on scaling
This is the best we have 
https://mooseframework.inl.gov/moose/source/systems/NonlinearSystemBase.html#scaling

The list of parameters of the executioner shows all the options you have. 
Sometimes scaling based on the residual instead of the Jacobian helps
Sometimes including off-diagonal terms also help, we've added that for scaling the Navier Stokes equations

**Reply by EdSheltonUKAEA** [reply]
Many thanks for all your help with this.  It looks like I need to go and do some reading to understand what might be causing my problem - and if possible fix it.
I need to take my understanding of the relevant mathematical processes applied in scaling or normalisation from the equivalent of A-level or simple undergraduate statistics in the UK (high school and college in the US?) to what MOOSE is doing when scaling solves to try and achieve an efficient and accurate solution.  I come from a materials science and engineering background, so not familiar with a lot of the applied math(s) and computer science terminology which is used.
Are there any books or other resources you can recommend?  How have others developed this knowledge, or does it usually just get learnt "on-the-job" while using the tools?

**Reply by GiudGiud** [reply]
@lindsayad do we have a more detailed explanation of what scaling does somewhere than the one I linked?

**Reply by lindsayad** [reply]
that is the best documentation

**Reply by lindsayad** [reply]
Regarding why you are seeing trouble with shorter time steps, it's possible that you are now resolving some physics that occurs on shorter time scales that you were not resolving before. I'm curious, have you tried running with no automatic scaling at all?

**Reply by EdSheltonUKAEA** [reply]
@lindsayad - Thanks.  Yes, when I previously discussed this problem with colleagues, turning off automatic scaling completely was the first thing they suggested.  It didn't fix the problem.

**Reply by lindsayad** [reply]
Ok good to know. The next thing I would explore is whether your Jacobian is accurate. Have you run any tests on small problems with `-snes_test_jacobian`?

**Reply by GiudGiud** [reply]
At this point, first I would exhaust the automatic scaling options (groups, off-diagonals, residual vs jacobian scaling)

Then I would try to replace each variable by a constant aux variable and see when good convergence is recovered. That could point to the troubled variable

**Reply by EdSheltonUKAEA** [reply]
I thought it worth giving an update on this  to let you know that I've (mostly) fixed the problem with my model.

It turns out it wasn't really to do with the scaling after all.  The problem was that I'd included unnecessary coupling of time derivatives between variables which weren't needed and so the problem was far more constrained than it needed to be and hence didn't work.  This arose from me misunderstanding what is meant by "coupling" with respect to different variables and kernels. 

Thanks for all your help answering this and my other recent questions which got my understanding to where it needed to be to have the eureka moment which solved the problem.
